### PR TITLE
Validated ODE integration

### PR DESCRIPTION
**Pull request by lbenet**

### CODE REVIEW

**Comment by lbenet** [comment]
cc: @dpsanders, @mforets, schillic

**Comment by dpsanders** [comment]
I'm getting
```
ERROR: LoadError: ArgumentError: Package TaylorModels does not have TaylorIntegration in its dependencies:
```

**Comment by dpsanders** [comment]
Sorry, just needed to `]resolve`.

**Comment by lbenet** [comment]
Shall I push a `Project.toml`; I guess this is the problem in travis...

**Comment by lbenet** [comment]
Tests pass again, at least in travis...

**Comment by lbenet** [comment]
Tests are failing for appveyor in x86; I truly think we should drop support for that.

**Comment by lbenet** [comment]
Rebasing to current master.

File: src/validatedODEs.jl [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            t0, tmax, orderT, abstol, parse_eqs)
+
+        # This is to compute dx[:][orderT] (now zero), needed for the remainder
+        f!(t, x, dx)
+
+        # Perform 25 checks of `check_property`; if the property is not satisfied
+        # throw an error
+        nsteps += 1
+        issatisfied = false
+        for nchecks = 1:25
```
Comment (schillic) [inline src/validatedODEs.jl]: Is there a fundamental reason for 25? Maybe make this a parameter.

File: src/validatedODEs.jl [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+
+            issatisfied = true
+            # New initial conditions
+            @inbounds for i in eachindex(x)
+                x[i]  = Taylor1( xTMN[i], orderT )
+                dx[i] = Taylor1( zero(xTMN[i]), orderT )
+            end
+            break
+        end
+        !issatisfied && error("`check_property is not satisfied: $t0 $(xv[nsteps])")
```
Comment (schillic) [inline src/validatedODEs.jl]:
```suggestion
        !issatisfied && error("`check_property` is not satisfied: $t0 $(xv[nsteps])")
```

File: src/validatedODEs.jl (hunk 1 above)
Comment (lbenet) [reply src/validatedODEs.jl]: Thanks for taking a look on this, @schillic !

No, there is actually no reason to have it 25. Yet, the heuristic argument is the following: 1/2^25 ≈ 3e-8, which is the smallest factor by which the original δt would be reduced, which seems reasonable. (To have it 30 means a factor close to 1.0e-9.)

 For the time being, this is purely experimental: we have to find out a good way to reduce time when it is needed. Experimenting with Laub-Loomis system has showed that it is not yet enough.

File: src/validatedODEs.jl (hunk 2 above)
Comment (lbenet) [reply src/validatedODEs.jl]: Nice catch! Thank you!

**Comment by lbenet** [comment]
I just pushed a new commit, where a new implementation (`validated_integ2`) was included. The design is rather different from the former one. It relies in a non-validated jet-transport integration, with proper bounds.

I would be thankful if you can perform independent tests.

**Comment by lbenet** [comment]
I pushed two new commits which fix the verificaation of contraction of the Picard iteration, and rewrite the `TaylorModelN` of each integration step, such that the actual remainder is absorved in the polynomial coefficients to be truly very small.

Question about the output: currently, the output of `validated_integ` is a vector with the evaluated times, and one with an `IntervalBox`es (flowpipe) containing the solution. I think it is better to have, instead of the actual polynomial instead of the flowpipe. Shall we get such a vector as output?

Except for the previous question, I think this is ready for review.

Cc: @dpsanders, @mforets, @schillic

**Comment by mforets** [comment]
Given that the algorithm produces the polynomial prior to overapproximation with an interval box, i agree that it's good to return it. Eventually, another tighter set representation can be used for the overapproximation.

**Comment by lbenet** [comment]
Thanks @mforets I'll experiment with it and try to come up with something which is not too complicate for plotting.

**Comment by mforets** [comment]
For example, for property checking one only needs a subset of the full flowpipe, so we can replace the evaluation of the bounding box over all variables, `xv[nsteps] = evaluate(xTMN, δq_norm)`, with the `min/max` bounds only on the variable of interest, to get better performance (particularly useful in higher dimensions) -- does it make sense?

**Comment by mforets** [comment]
If an equation is of the form `dx[i] = c`, where `c::Float64`, i get errors of the form

```julia
ERROR: LoadError: BoundsError: attempt to access 1-element Array{TaylorN{Float64},1} at index [2]
Stacktrace:
 [1] getindex at ./array.jl:729 [inlined]
 [2] getindex at /Users/forets/.julia/dev/TaylorSeries/src/auxiliary.jl:76 [inlined]
 [3] jetcoeffs!(::getfield(Main, Symbol("##21#23")), ::Taylor1{Float64}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Tay

lor1{TaylorN{Float64}},1}) at /Users/forets/.julia/dev/TaylorIntegration/src/explicitode.jl:79
 [4] __jetcoeffs!(::Val{false}, ::Function, ::Taylor1{Float64}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{Tayl
orN{Float64}},1}) at /Users/forets/.julia/dev/TaylorIntegration/src/explicitode.jl:96
 [5] taylorstep!(::Function, ::Taylor1{Float64}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{TaylorN{Float64}},1
}, ::Float64, ::Float64, ::Array{TaylorN{Float64},1}, ::Int64, ::Float64, ::Bool) at /Users/forets/.julia/dev/TaylorIntegration/src/explicitode.jl:201
 [6] validated_step!(::getfield(Main, Symbol("##21#23")), ::Taylor1{Float64}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Array{Taylor1{TaylorN{Float64}},1}, ::Arra
y{Taylor1{TaylorN{Float64}},1}, ::Taylor1{Float64}, ::Array{Taylor1{IntervalArithmetic.Interval{Float64}},1}, ::Array{Taylor1{IntervalArithmetic.Interval{Float6
4}},1}, ::Array{Taylor1{IntervalArithmetic.Interval{Float64}},1}, ::Float64, ::Float64, ::Array{TaylorN{Float64},1}, ::Array{IntervalArithmetic.Interval{Float64
},1}, ::Array{TaylorModels.TaylorModelN{5,Float64,Float64},1}, ::Array{IntervalArithmetic.IntervalBox{5,Float64},1}, ::Array{IntervalArithmetic.Interval{Float64
},1}, ::IntervalArithmetic.IntervalBox{5,Float64}, ::IntervalArithmetic.IntervalBox{5,Float64}, ::IntervalArithmetic.IntervalBox{5,Float64}, ::Int64, ::Int64, :
:Float64, ::Bool, ::getfield(Reachability.ReachSets, Symbol("##61#62"))) at /Users/forets/.julia/packages/TaylorModels/79LT9/src/validatedODEs.jl:139
 [7] #validated_integ#31(::Int64, ::Bool, ::Function, ::Function, ::Function, ::Array{Float64,1}, ::IntervalArithmetic.IntervalBox{5,Float64}, ::Float64, ::Floa
t64, ::Int64, ::Int64, ::Float64) at /Users/forets/.julia/packages/TaylorModels/79LT9/src/validatedODEs.jl:268
 [8] (::getfield(TaylorModels, Symbol("#kw##validated_integ")))(::NamedTuple{(:maxsteps, :check_property),Tuple{Int64,getfield(Reachability.ReachSets, Symbol("#
#61#62"))}}, ::typeof(TaylorModels.validated_integ), ::Function, ::Array{Float64,1}, ::IntervalArithmetic.IntervalBox{5,Float64}, ::Float64, ::Float64, ::Int64,
 ::Int64, ::Float64) at ./none:0
```

I think this may be fixable here, by converting the right-hand side to the appropriate type.

For an example see: https://github.com/mforets/escritoire/blob/master/reachability/BenchsHybrid.ipynb
changing `dx[2]` to `dx[2] = -1.0` in  the function`bball!(t, x, dx)`.

**Comment by lbenet** [comment]
Thanks for reporting; the issue is actually in `TaylorIntegration.jl` and is related with promotion of `TaylorSeries`. The simplest cure is to add `zero(x[1])`, which should yield the correct `Taylor1` type. The subtlety of the problem is that promotion knows nothing about the `order` of the Taylor polynomial, which complicaates things.

**Comment by lbenet** [comment]
I'm merging this to make it available in master. It currently works, though there are many itchy corners. Any further development may continue in another PR.

**Pull request merged**